src/openlcb/IfTcpImpl.hxx

```diff
@@ -0,0 +1,553 @@
+        {
+            HASSERT(bufOfs_ >= bufEnd_);
+            return call_immediately(STATE(read_more_bytes));
+        }
+    }
+
+    Action read_more_bytes()
+    {
+        if (bufEnd_ <= bufOfs_)
+        {
+            bufOfs_ = 0;
+            bufEnd_ = 0;
+        }
+        return read_single(&helper_, device()->fd(), buffer_ + bufEnd_,
+            READ_BUFFER_SIZE - bufEnd_, STATE(read_done), READ_PRIO);
+    }
+
+    Action read_done()
+    {
+        if (helper_.hasError_)
+        {
+            notify_barrier();
+            set_terminated();
+            device()->report_read_error();
+            return exit();
+        }
+        bufEnd_ = READ_BUFFER_SIZE - helper_.remaining_;
+        return parse_bytes();
+    }
+
+    Action send_entry()
+    {
+        return allocate_and_call(dst_, STATE(have_alloc_msg));
+    }
+
+    Action have_alloc_msg()
+    {
+        auto *b = get_allocation_result(dst_);
+        auto prio = TcpDefs::guess_priority(msg_);
+        b->data()->assign(std::move(msg_));
+        b->data()->skipMember_ = skipMember_;
+        /// @todo (balazs.racz): there should be some form of throttling here,
+        /// ot not read more bytes from the tcp socket than how much RAM we
+        /// have available.
+        dst_->send(b, prio);
+        return call_immediately(STATE(start_msg));
+    }
+
+    /** Calls into the parent flow's barrier notify, but makes sure to
+     * only do this once in the lifetime of *this. */
+    void notify_barrier()
+    {
+        if (barrierOwned_)
+        {
+            barrierOwned_ = false;
+            device()->barrier_.notify();
+        }
+    }
+
+    /// We attempt to read this many bytes in one go from the FD.
+    static constexpr unsigned READ_BUFFER_SIZE = 300;
+    /// If we have to guess at the size of the packet, we start by allocating
+    /// this many bytes.
+    static constexpr unsigned DEFAULT_PACKET_SIZE =
+        TcpDefs::MIN_ADR_MESSAGE_SIZE + 16;
+    static constexpr unsigned MIN_SIZE_GUESS = TcpDefs::HDR_SIZE_END;
+    /// What priority to use for reads from fds.
+    static constexpr unsigned READ_PRIO = Selectable::MAX_PRIO;
+
+    /// Temporary buffer to read into from the FD.
+    uint8_t buffer_[READ_BUFFER_SIZE];
+    /// true iff pending parent->barrier_.notify()
+    uint8_t barrierOwned_{1};
+    /// Offset of the end in the read buffer.
+    uint16_t bufEnd_;
+    /// First active byte (offset of the beginning) in the read buffer.
+    uint16_t bufOfs_;
+    /// How many bytes we think the current message will be. -1 if we don't
+    /// know yet.
+    int expectedLen_;
+    /// Assembly buffer. Holds the captured message so far.
+    string msg_;
+    /// Where to send parsed messages to.
+    HubPortInterface *dst_;
+    /// Parsed messages will be initialized to this skipMember_.
+    HubPortInterface *skipMember_;
+    StateFlowSelectHelper helper_{this};
+};
+
+using TcpHubDeviceSelect = HubDeviceSelect<HubFlow, FdToTcpParser>;
```

Comment:
Is it good practice to have "using" in the global scope of a header file?

Reply:
using namespace foobar; is forbidden.
the syntax here is a typedef, just formatted as a more readable equivalent since c++11.